File: backend/src/main/java/com/backend/api/search/service/GlobalSearchService.java
package com.backend.api.search.service;

import com.backend.api.search.dto.SearchPageResponse;
import com.backend.api.search.dto.SearchResultDto;
import com.backend.api.user.service.UserSearchService;
import com.backend.domain.user.entity.search.UserDocument;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class GlobalSearchService {

    private final UserSearchService userSearchService;
    // private final PostSearchService postSearchService;
    // private final QuestionSearchService questionSearchService;

    //유저 검색
    public SearchPageResponse<SearchResultDto> searchUser(String keyword, int page, int size) {
        Page<UserDocument> users = userSearchService.search(keyword, page, size);

        List<SearchResultDto> content = users.getContent().stream()
                .map(u -> SearchResultDto.builder()
                        .type("user")
                        .id(u.getId())
                        .title(u.getName())
                        .snippet(u.getNickname())
                        .build())
                .collect(Collectors.toList());

        return SearchPageResponse.from(users, content);
    }

    //통합검색 (나중에 Post, Question까지 합칠 예정)

    public SearchPageResponse<SearchResultDto> searchAll(String keyword, int page, int size) {
        // 지금은 유저 결과만 반환
        return searchUser(keyword, page, size);


        /*
        List<SearchResultDto> merged = new ArrayList<>();
        merged.addAll(searchUser(keyword, page, size).getContent());
        merged.addAll(searchPost(keyword, page, size).getContent());
        merged.addAll(searchQuestion(keyword, page, size).getContent());

        return SearchPageResponse.<SearchResultDto>builder()
                .content(merged)
                .currentPage(page)
                .totalElements(merged.size())
                .totalPages(1)
                .last(true)
                .build();
        */
    }

    /*
    public SearchPageResponse<SearchResultDto> searchPost(String keyword, int page, int size) {
        Page<PostDocument> posts = postSearchService.search(keyword, page, size);
        List<SearchResultDto> content = posts.getContent().stream()
                .map(p -> SearchResultDto.builder()
                        .type("post")
                        .id(p.getId())
                        .title(p.getTitle())
                        .snippet(p.getContent())
                        .build())
                .toList();

        return SearchPageResponse.from(posts, content);
    }

    public SearchPageResponse<SearchResultDto> searchQuestion(String keyword, int page, int size) {
        Page<QuestionDocument> questions = questionSearchService.search(keyword, page, size);
        List<SearchResultDto> content = questions.getContent().stream()
                .map(q -> SearchResultDto.builder()
                        .type("question")
                        .id(q.getId())
                        .title(q.getTitle())
                        .snippet(q.getBody())
                        .build())
                .toList();

        return SearchPageResponse.from(questions, content);
    }
    */
}

File: backend/src/main/java/com/backend/api/search/service/GlobalSearchService.kt
package com.backend.api.search.service

import com.backend.api.search.dto.SearchPageResponse
import com.backend.api.search.dto.SearchResultDto
import com.backend.api.user.service.UserSearchService
import com.backend.domain.user.entity.search.UserDocument
import org.springframework.data.domain.Page
import org.springframework.stereotype.Service

@Service
class GlobalSearchService(
    private val userSearchService: UserSearchService
    // private val postSearchService: PostSearchService
    // private val questionSearchService: QuestionSearchService
) {

    // 유저 검색 (Elasticsearch 기반)
    fun searchUser(keyword: String, page: Int, size: Int): SearchPageResponse<SearchResultDto> {
        val users: Page<UserDocument> = userSearchService.search(keyword, page, size)

        val content = users.content.map { u ->
            SearchResultDto(
                type = "user",
                id = u.id ?: "",
                title = u.name ?: "",
                snippet = u.nickname ?: ""
            )
        }

        return SearchPageResponse.from(users, content)
    }

    // 통합 검색 (현재는 User만)
    fun searchAll(keyword: String, page: Int, size: Int): SearchPageResponse<SearchResultDto> {
        return searchUser(keyword, page, size)
    }

    /*
    fun searchPost(keyword: String, page: Int, size: Int): SearchPageResponse<SearchResultDto> {
        val posts = postSearchService.search(keyword, page, size)
        val content = posts.content.map { p ->
            SearchResultDto(
                type = "post",
                id = p.id,
                title = p.title,
                snippet = p.content
            )
        }
        return SearchPageResponse.from(posts, content)
    }

    fun searchQuestion(keyword: String, page: Int, size: Int): SearchPageResponse<SearchResultDto> {
        val questions = questionSearchService.search(keyword, page, size)
        val content = questions.content.map { q ->
            SearchResultDto(
                type = "question",
                id = q.id,
                title = q.title,
                snippet = q.body
            )
        }
        return SearchPageResponse.from(questions, content)
    }
    */
}
